core/src/main/scala/org/apache/spark/storage/BlockManagerMasterEndpoint.scala

```diff
@@ -657,6 +688,38 @@ class BlockManagerMasterEndpoint(
     }
   }
 
+  private def getShufflePushMergerLocations(
+      numMergersNeeded: Int,
+      hostsToFilter: Set[String]): Seq[BlockManagerId] = {
+    val blockManagersWithExecutors = blockManagerIdByExecutor.groupBy(_._2.host)
+      .mapValues(_.head).values.map(_._2).toSet
+    val filteredBlockManagersWithExecutors = blockManagersWithExecutors
+      .filterNot(x => hostsToFilter.contains(x.host))
+    val filteredMergersWithExecutors = filteredBlockManagersWithExecutors.map(
+      x => BlockManagerId(x.executorId, x.host, StorageUtils.externalShuffleServicePort(conf)))
+
+    // Enough mergers are available as part of active executors list
+    if (filteredMergersWithExecutors.size >= numMergersNeeded) {
+      filteredMergersWithExecutors.toSeq
+    } else {
+      // Delta mergers added from inactive mergers list to the active mergers list
+      val filteredMergersWithExecutorsHosts = filteredMergersWithExecutors.map(_.host)
+      // Pick random hosts instead of preferring the top of the list
+      val randomizedShuffleMergerLocations = Utils.randomize(shuffleMergerLocations.values.toSeq)
```

Comment:
Shall we randomize the filtered locations instead? Thus, we could randomize fewer locations.

Reply:
If all of the filtered locations are needed then we pretty much we won't have any shuffling right? Since we don't know how many more we need I think we have to randomize the `shuffleMergerLocations` list. Thoughts?

Reply:
If I am not wrong, @Ngone51's suggestion is following:
```
// remove randomizedShuffleMergerLocations

val filteredMergersWithoutExecutors = filteredMergersWithExecutorsHosts 
        .filterNot(x => hostsToFilter.contains(x.host))
        .filterNot(x => filteredMergersWithExecutorsHosts.contains(x.host))
// We can skip randomize if filteredMergersWithoutExecutors.size <= numMergersNeeded - filteredMergersWithExecutors.size
val randomizedFilteredLocations = Utils.randomize(filteredMergersWithoutExecutors)

return  filteredMergersWithExecutors.toSeq ++ randomizedFilteredLocations
        .take(numMergersNeeded - filteredMergersWithExecutors.size)
```

Reply:
Yes! Thanks @mridulm

Reply:
@mridulm shouldn't it be the following instead?
```
val filteredMergersWithoutExecutors = shuffleMergerLocations.values.toSeq
  .filterNot(....)
  .filterNot(....)
....
```

Reply:
@Victsm You're right @venkata91 fixed in the right way.